core/stop/residual_norm.cpp

```diff
@@ -234,6 +236,212 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_TYPE(GKO_DECLARE_RESIDUAL_NORM);
     class ImplicitResidualNorm<_type>
 GKO_INSTANTIATE_FOR_EACH_VALUE_TYPE(GKO_DECLARE_IMPLICIT_RESIDUAL_NORM);
 
+class ResidualNormFactory;
+
+struct residual_norm_factory_parameters
+    : public enable_parameters_type<residual_norm_factory_parameters,
+                                    ResidualNormFactory> {
+    double GKO_FACTORY_PARAMETER_SCALAR(threshold, 0.0);
+
+    mode GKO_FACTORY_PARAMETER_SCALAR(baseline, mode::rhs_norm);
+};
+
+
+class ResidualNormFactory
+    : public EnablePolymorphicObject<ResidualNormFactory, CriterionFactory>,
+      public EnablePolymorphicAssignment<ResidualNormFactory> {
+    friend class EnablePolymorphicObject<ResidualNormFactory, CriterionFactory>;
+    friend class enable_parameters_type<residual_norm_factory_parameters,
+                                        ResidualNormFactory>;
+    friend EnableDefaultCriterionFactory<ResidualNormFactory, Criterion,
+                                         residual_norm_factory_parameters>;
+
+    explicit ResidualNormFactory(
+        std::shared_ptr<const Executor> exec,
+        const residual_norm_factory_parameters& parameters = {})
+        : EnablePolymorphicObject<ResidualNormFactory, CriterionFactory>(
+              std::move(exec)),
+          parameters_{parameters}
+    {}
+
+    std::unique_ptr<Criterion> generate_impl(CriterionArgs args) const override
+    {
+        std::unique_ptr<Criterion> result;
+        auto exec = this->get_executor();
+        run<matrix::Dense<double>, matrix::Dense<float>, matrix::Dense<half>,
+            matrix::Dense<bfloat16>, matrix::Dense<std::complex<double>>,
+            matrix::Dense<std::complex<float>>,
+            matrix::Dense<std::complex<half>>,
+            matrix::Dense<std::complex<bfloat16>>
```

Comment:
need to adapt with bfloat16 and half option. I think it is the reason for failed CI test